### DISCUSSION TITLE
is there a way to modify the context for a specific operator?

### DISCUSSION THREAD

**Discussion by karakanb** [apache/airflow#34449] [Q&A]
I have an operator that I'd like to enable Jinja, but not give access to the whole context. I'd like to create my own set of variables that are available to the templates, and I imagine by overriding a certain function I should be able to enable that behavior. Is there a way to do that? I'd appreciate any pointers around where exactly the rendering of the templated fields happens in the code.

**Comment by potiuk** [comment]
You can use your own Macros: https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#macros and expose them to the operators via plugins mechanism: https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/plugins.html

You cannot dynamically remove or change context per operator call - this has to be done via macros and plugins - you make macros available in `macros` context variable.

This is basic assumption that operators have access to the full context when they are running and it cannot be changed to provide "subset" of context. 

You have not explained why you want to do it, so I am just guessing why you would like to do it.

If you want to treat as security feature, it's not one. Each operator is python code that - no matter what you do - has access to everything that is available in worker task and there is no way to sandbox or block it. All information in context is (and will be) always accessible to the operator's python code, so if your goal is security, this is wrong way because it gives you false sense of security. If you are worried about what DAG authors can do you have other ways as Deployment Manager to limit those (code reviews, static code checks before submitting DAGs etc.).

See Security Model of Airflow for details:

https://airflow.apache.org/docs/apache-airflow/stable/security/security_model.html

**Reply by karakanb** [reply]
Thanks a lot for the comment. My goal is not to secure the operator code, but more what the template can do. Let me give you an example:
- I allow people to set up an S3 sensor via a web UI where they put in the configuration and I save it to a database.
- I take that configuration and dynamically build pipelines on Airflow.
- When the sensor runs, I'd like the `bucket_key` to be templated; however, I don't want `conn.xyz.login` in that template to work, only the `ds`, for instance.
- Therefore, I'd like to build an operator that overrides what's available in the templating context.

In this sense, the operator code is trusted already, and I want the operator class to have access to everything, but I want to be able to override a single function and modify the available context for the templated fields, so that `bucket_key` in the example above can only access a specific set of values rather than the full Airflow context. 

Does it make sense?
